File: week2/secp256k1/python/jacobian.py
from typing import Optional, Tuple
from affine import affine_add, inv, P

JacPoint = Tuple[int, int, int]
AffPoint = Optional[Tuple[int, int]]


def to_affine(p: JacPoint) -> AffPoint:
    """Convert Jacobian point (X:Y:Z) to affine coordinates (x, y)."""
    X, Y, Z = p
    if Z == 0:
        return None
    z_inv = inv(Z)
    z2 = (z_inv * z_inv) % P
    z3 = (z2 * z_inv) % P
    return (X * z2 % P, Y * z3 % P)


def jacobian_double(p: JacPoint) -> JacPoint:
    """Double a point in Jacobian coordinates."""
    X1, Y1, Z1 = p
    if Y1 == 0:
        return (0, 1, 0)  # Point at infinity

    Y1_sq = Y1 * Y1 % P
    Y1_4 = Y1_sq * Y1_sq % P
    S = 4 * X1 * Y1_sq % P
    M = 3 * X1 * X1 % P
    X3 = (M * M - 2 * S) % P
    Y3 = (M * (S - X3) - 8 * Y1_4) % P
    Z3 = (2 * Y1 * Z1) % P
    return (X3, Y3, Z3)


def jacobian_add_jac(p: JacPoint, q: JacPoint) -> JacPoint:
    """Add two Jacobian points, returning a Jacobian point."""
    X1, Y1, Z1 = p
    X2, Y2, Z2 = q

    if Z1 == 0:
        return q
    if Z2 == 0:
        return p

    Z1_sq = Z1 * Z1 % P
    Z2_sq = Z2 * Z2 % P
    Z1_cu = Z1_sq * Z1 % P
    Z2_cu = Z2_sq * Z2 % P

    U1 = X1 * Z2_sq % P
    U2 = X2 * Z1_sq % P
    S1 = Y1 * Z2_cu % P
    S2 = Y2 * Z1_cu % P

    if U1 == U2:
        if S1 != S2:
            return (0, 1, 0)  # Point at infinity
        return jacobian_double(p)

    H = (U2 - U1) % P
    R = (S2 - S1) % P
    H_sq = H * H % P
    H_cu = H_sq * H % P
    U1_H_sq = U1 * H_sq % P

    X3 = (R * R - H_cu - 2 * U1_H_sq) % P
    Y3 = (R * (U1_H_sq - X3) - S1 * H_cu) % P
    Z3 = Z1 * Z2 * H % P

    return (X3, Y3, Z3)


def jacobian_add(p: JacPoint, q: JacPoint) -> AffPoint:
    """Add two Jacobian points, returning an affine point."""
    return to_affine(jacobian_add_jac(p, q))

File: week2/secp256k1/rust/src/jacobian.rs
use crate::affine::{mod_add, mod_inv, mod_mul, mod_sub};
use num_bigint::BigUint;
use num_traits::Zero;

/// Convert Jacobian (X,Y,Z) to affine (x,y)
pub fn to_affine(j: &(BigUint, BigUint, BigUint)) -> Option<(BigUint, BigUint)> {
    let (x, y, z) = j;
    if z.is_zero() {
        return None;
    }
    let z_inv = mod_inv(z);
    let z2 = mod_mul(&z_inv, &z_inv);
    let z3 = mod_mul(&z2, &z_inv);
    Some((mod_mul(x, &z2), mod_mul(y, &z3)))
}

/// Jacobian point addition: P + Q = R (converted to affine)
pub fn jacobian_add(
    p: &(BigUint, BigUint, BigUint),
    q: &(BigUint, BigUint, BigUint),
) -> Option<(BigUint, BigUint)> {
    let (x1, y1, z1) = p;
    let (x2, y2, z2) = q;

    if z1.is_zero() {
        return to_affine(q);
    }
    if z2.is_zero() {
        return to_affine(p);
    }

    let z1z1 = mod_mul(z1, z1);
    let z2z2 = mod_mul(z2, z2);

    let u1 = mod_mul(x1, &z2z2);
    let u2 = mod_mul(x2, &z1z1);

    let z1z1z1 = mod_mul(&z1z1, z1);
    let z2z2z2 = mod_mul(&z2z2, z2);

    let s1 = mod_mul(y1, &z2z2z2);
    let s2 = mod_mul(y2, &z1z1z1);

    if u1 == u2 {
        if s1 != s2 {
            return None; // P + (-P) = ∞
        }
        return jacobian_double(p);
    }

    let h = mod_sub(&u2, &u1);
    let r = mod_sub(&s2, &s1);

    let h2 = mod_mul(&h, &h);
    let h3 = mod_mul(&h2, &h);
    let u1h2 = mod_mul(&u1, &h2);

    let x3 = mod_sub(&mod_sub(&mod_mul(&r, &r), &h3), &mod_add(&u1h2, &u1h2));
    let y3 = mod_sub(&mod_mul(&r, &mod_sub(&u1h2, &x3)), &mod_mul(&s1, &h3));
    let z3 = mod_mul(&mod_mul(z1, z2), &h);

    to_affine(&(x3, y3, z3))
}

/// Jacobian point doubling: R = 2P (converted to affine)
pub fn jacobian_double(p: &(BigUint, BigUint, BigUint)) -> Option<(BigUint, BigUint)> {
    let (x, y, z) = p;

    if y.is_zero() {
        return None; // tangent is vertical ⇒ point at infinity
    }

    let yy = mod_mul(y, y);
    let yyyy = mod_mul(&yy, &yy);
    let zz = mod_mul(z, z);

    let s = mod_mul(&mod_mul(&BigUint::from(4u32), x), &yy);
    let m = mod_add(
        &mod_mul(&BigUint::from(3u32), &mod_mul(x, x)),
        &mod_mul(&BigUint::from(0u32), &mod_mul(&zz, &zz)), // a = 0 in secp256k1
    );

    let x3 = mod_sub(&mod_mul(&m, &m), &mod_add(&s, &s));
    let y3 = mod_sub(
        &mod_mul(&m, &mod_sub(&s, &x3)),
        &mod_mul(&BigUint::from(8u32), &yyyy),
    );
    let z3 = mod_mul(y, z).clone();
    let z3 = mod_mul(&BigUint::from(2u32), &z3);

    to_affine(&(x3, y3, z3))
}
